resolves #925

beratbayram opened the pull request:
Fixes #925.

WHAT: It changes the output of next.js with TS template.
WHY: Basic templating is broken.
HOW: Existing customers are not affected. New customers will have a better experience.

---

- Tests
  - [ ] TODO (before merge)
  - [ ] Tests included
  - [x] Tests not necessary because: already written
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required
  - [ ] Not required because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <!--e.g. <https://github.com/cloudflare/cloudflare-docs/pull/>...-->
  - [x] Documentation not necessary because: Changes the output not the usage

---

I've run tests and two fails as follows but I don't believe they are related to my changes.

<img width="678" alt="image" src="https://github.com/user-attachments/assets/83f0e811-0e40-496f-a9b8-44d25af4cc5b" />

<img width="798" alt="image" src="https://github.com/user-attachments/assets/c5f2bbb0-bba3-4f54-9b6d-82d033acd1e3" />

beratbayram commented:
[dario-piotrowicz](https://github.com/dario-piotrowicz)

vicb commented on packages/create-cloudflare/templates/next/c3.ts:
```diff
@@ -31,29 +32,55 @@ const generate = async (ctx: C3Context) => {
+	// use bindings during local development (when running the application with \`next dev\`).
+	// For more information see:
+	// https://github.com/cloudflare/next-on-pages/blob/main/internal-packages/next-dev/README.md
+	if (process.env.NODE_ENV === 'development') {
```
@dario-piotrowicz do we need `(process.env.NODE_ENV === 'development')` here? (or could it be inside `setupDevPlatform` ?

(The code here is only moved from l42 so not directly related to the PR)

dario-piotrowicz replied:
This check is totally unnecessary, moving it inside `setupDevPlatform` wouldn't actually accomplish anything

it's here to try to make it extra clear to the developer that this function is only needed in development, in reality this only has any effect during development anyways

I'm happy to get rid of it if you want 👍

vicb replied:
> This check is totally unnecessary

Then I would drop it.

And in [next-on-pages](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/pages-e2e/fixtures/issue797app/next.config.mjs#L6) as well

vicb commented on .changeset/moody-walls-shake.md:
```diff
@@ -0,0 +1,9 @@
+---
+"create-cloudflare": major
```
probably only a patch? (bug fix)

vicb replied:
> I've run pnpm changeset, and it decided that it is a major change

I have to admit that the "UI" is weird but `changeset` does not decide anything for you, it only depends on what you type :)

beratbayram replied:
> changeset does not decide anything for you

I had thought there was some logic behind it. My bad. I've just changed it to patch as you suggested

vicb reviewed (commented):
I took a quick look here but I don't really think the changes are optimal, there should be no reason to need both a TS and JS version of `updateNextXConfig`.

One thing that look weird is that we do not pass `--js` or `--ts` to `create-next-app
`https://github.com/cloudflare/workers-sdk/blob/6c2e94318dd079f9cba3f763e67ba6698ef597f6/packages/create-cloudflare/templates/next/c3.ts#L27

beratbayram commented:
> I took a quick look here but I don't really think the changes are optimal, there should be no reason to need both a TS and JS version of `updateNextXConfig`.

JS version only prepends some code to existing config while TS version removes the existing one completely and creates a new hardcoded one from scratch. I can make JS version also behave like TS version but in case of a possible config change by next.js team, both versions will be affected instead of only the TS version.

> One thing that look weird is that we do not pass `--js` or `--ts` to `create-next-app `
> 
> https://github.com/cloudflare/workers-sdk/blob/6c2e94318dd079f9cba3f763e67ba6698ef597f6/packages/create-cloudflare/templates/next/c3.ts#L27

We don't need that, user decides that during next-cli prompt

vicb commented:
> We don't need that, user decides that during next-cli prompt

Thanks for the clarification here.

> JS version only prepends some code to existing config while TS version removes the existing one completely and creates a new hardcoded one from scratch. I can make JS version also behave like TS version but in case of a possible config change by next.js team, both versions will be affected instead of only the TS version.

Could you expand on why the TS config would need to be fully replaced? (maybe add the generated TS config + explanation to the PR description). Thanks!

beratbayram commented:
> Thanks for the clarification here.

No problem

> Could you expand on why the TS config would need to be fully replaced? (maybe add the generated TS config + explanation to the PR description). Thanks!

I've just edited the description. Please also check the conversation at the linked issue.

dario-piotrowicz commented:
> > because C3 generates an await statement inside a .ts file which is illegal
> 
> I think we should drop the `await`, see:
> 
> [cloudflare/next-on-pages@`2cd4c3c`/internal-packages/next-dev/src/index.ts#L6-L13](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/src/index.ts#L6-L13)
> 
> [cloudflare/next-on-pages@`2cd4c3c`/internal-packages/next-dev/README.md?plain=1#L56-L59](https://github.com/cloudflare/next-on-pages/blob/2cd4c3c704a00e6b693229f1f14102abc6318d11/internal-packages/next-dev/README.md?plain=1#L56-L59)
> 
> @dario-piotrowicz what do you think?

yeah I forgot about the `await`ing not being actually necessary, yes probably removing it would be the simplest thing, I feel like awaiting the function provides a nice safety net (since we've seen that not awaiting the function is totally fine, but there is no hard guarantee about that, in theory it could no longer be true if Next.js changed how they run their dev server etc...), but since now it complicates the C3 logic here I think we might just get rid of it

beratbayram commented:
I've reverted all of my previous work.

> yeah I forgot about the awaiting not being actually necessary, yes probably removing it would be the simplest thing,

Removed `await` and added `.catch(console.error)`

> it's here to try to make it extra clear to the developer that this function is only needed in development

Removed the if and improved the comment to emphasize that.

vicb commented on .changeset/famous-coats-admire.md:
```diff
@@ -0,0 +1,7 @@
+"create-cloudflare": patch
+---
+
+# \[create-cloudflare\] next.js with ts fix #7854
```
```suggestion
Update the Next.js template
```

`#` will generate a weird formatting in the changelog

vicb commented on .changeset/famous-coats-admire.md:
```diff
@@ -0,0 +1,7 @@
+
+# \[create-cloudflare\] next.js with ts fix #7854
+
+Removed top-level await in next.config.ts
```
```suggestion
Removed top-level `await` in `next.config.ts` as it is not allowed there and not required.
```

vicb reviewed (approved):
LGTM, thanks for the PR and the updates!

I triggered the e2e

vicb commented on .changeset/famous-coats-admire.md:
```diff
@@ -0,0 +1,8 @@
+Update the Next.js template
+
+* Removed top-level `await` in `next.config.ts` as it is not allowed there and not required.
+* Improved setupDevPlatform() comment
```
well then:

```suggestion
* Improved `setupDevPlatform()` comment
```

beratbayram commented:
Thanks for your directions and your patience. It was fun to work on this.

vicb commented:
@beratbayram it looks like the formatting is wrong, could you please run `pnpm fix` on the change?

beratbayram commented:
> @beratbayram it looks like the formatting is wrong, could you please run `pnpm fix` on the change?

done
